Source/MLX/MLXArray.swift

```diff
@@ -246,6 +241,109 @@ public final class MLXArray {
+            .max()
+            ?? self.size
+    }
+
+    func copy(from: UnsafeRawBufferPointer, to output: UnsafeMutableRawBufferPointer) {
+        let contiguousDimension = self.contiguousToDimension()
+
+        if contiguousDimension == 0 {
+            // entire backing is contiguous
+            from.copyBytes(to: output)
+
+        } else {
+            // only part of the backing is contiguous (possibly a single element)
+            // iterate the non-contiguous parts and copy the contiguous chunks into
+            // the output.
+
+            // these are the parts to iterate
+            let shape = self.shape.prefix(upTo: contiguousDimension)
+            let strides = self.strides.prefix(upTo: contiguousDimension)
+            let ndim = contiguousDimension
+            let itemSize = self.itemSize
+
+            // the size of each chunk that we copy.  this computes the stride of
+            // (contiguousDimension - 1) if it were contiguous
+            let destItemSize: Int
+            if contiguousDimension == self.ndim {
+                // nothing contiguous
+                destItemSize = itemSize
+            } else {
+                destItemSize =
+                    self.strides[contiguousDimension] * self.shape[contiguousDimension] * itemSize
+            }
+
+            // the index of the current source item
+            var index = Array.init(repeating: 0, count: ndim)
+
+            // output pointer
+            var dest = output.baseAddress!
+
+            while true {
+                // compute the source index by multiplying the index by the
+                // stride for each dimension
+
+                // note: in the case where the array has negative strides / offset
+                // the base pointer we have will have the offset already applied,
+                // e.g. asStrided(a, [3, 3], strides: [-3, -1], offset: 8)
+
+                let sourceIndex = zip(index, strides).reduce(0) { $0 + ($1.0 * $1.1) }
+
+                // convert to byte pointer
+                let src = from.baseAddress! + sourceIndex * itemSize
+                dest.copyMemory(from: src, byteCount: destItemSize)
+
+                // next output address
+                dest += destItemSize
+
+                // increment the index
+                for dimension in Swift.stride(from: ndim - 1, through: 0, by: -1) {
+                    // do we need to "carry" into the next dimension?
+                    if index[dimension] == (shape[dimension] - 1) {
+                        if dimension == 0 {
+                            // all done
+                            return
+                        }
+
+                        index[dimension] = 0
+                    } else {
+                        // just increment the dimension and we are done
+                        index[dimension] += 1
+                        break
+                    }
+                }
+            }
+
+        }
+    }
```

Comment:
This chunk is dealing with non-contiguous memory and being able to copy it out into contiguous memory.  See also the tests.